db/migrations/20201020182259_backfill_kpack_lifecycle_data_association_to_droplets.rb

```diff
@@ -0,0 +1,18 @@
+Sequel.migration do
+  up do
+    unassociated_kpack_lifecycle_data = self[:kpack_lifecycle_data].select(:id, :build_guid).where(droplet_guid: nil).exclude(build_guid: nil)
+
+    unassociated_kpack_lifecycle_data.each do |kld|
+      droplet = self[:droplets].select(:guid).where(build_guid: kld[:build_guid]).first
+
+      # there may not have been a droplet created yet for the build
+      next if droplet.nil?
```

Comment:
Yeah I think that's the same as what I mentioned in this comment:
https://github.com/cloudfoundry/cloud_controller_ng/pull/1917#discussion_r508857218

I'm not sure if there is a way around it apart from either:

* Reverting 9176948 and always supporting the ability to update image for "docker" and "kpack" droplets.
* Running the migration again as part of a future release (this requires people to upgrade to this version first -- if they jump and skip this version then same problem)
* A Kubernetes job that runs on every cf-for-k8s upgrade and migrates them which we eventually delete? 🤔
* We accept it as an edge case and document the workaround (repush) if stack updates fail for an app

Reply:
> I think there is an unlikely (?) race condition here where the ccdb-migrate job runs this but there are still cf-api-servers out there with the old code making "docker" droplets. If there are any apps in this state, they won't be able to get stack updates until the app is repushed (repushing is a fairly easy mitigation IMO).

> Scenario I'm imagining would require one cf-api-server pod to rollout with new changes and one to remain with old changes. cf push gets its request routed to the old one => droplet eventually created without associated kpack_lifecycle_data

This is almost certainly possible, and was not something I had considered when Tim and I talked about this earlier. One thing that we've never been able to build on BOSH that would totally fix this type of deploy scenario is a post-deploy migration job in addition to the pre-deploy one. If we had that running its own, idempotent post-deploy migrations, we could clean up after ourselves in scenarios like this and maybe even use it implement proper [expand-and-contract parallel-change migrations](https://martinfowler.com/articles/evodb.html).

> We accept it as an edge case and document the workaround (repush) if stack updates fail for an app

I think we're still early enough in the life of cf-for-k8s that this is acceptable, particularly if you consider this relative to how broken other things are.

Reply:
> I think we're still early enough in the life of cf-for-k8s that this is acceptable, particularly if you consider this relative to how broken other things are

I think it may be cheaper to just leave this in now versus doing it later and risking the context being lost. 🤷 

Cause of issues like https://github.com/cloudfoundry/capi-k8s-release/issues/88 where we're literally missing data on some of the droplets, I think folks may have to repush anyway.

Reply:
Actually, as I think about this more... we might be able to use the fact that historical kpack-built droplets are "docker" to help resolve that issue (https://github.com/cloudfoundry/capi-k8s-release/issues/88).

It will be difficult to know when it is safe for a "kpack" droplet to use `@process.started_command` if some "kpack" droplets have the incomplete command, but if we fix that issue at the same time as we introduce "kpack" droplets then we can assume that all of them have complete detected start commands. 🤔

Reply:
> It will be difficult to know when it is safe for a "kpack" droplet to use @process.started_command if some "kpack" droplets have the incomplete command, but if we fix that issue at the same time as we introduce "kpack" droplets then we can assume that all of them have complete detected start commands. 🤔

This is a good point. It's a kinda risky two-birds-with-one-stone way of fixing these bugs, but it sounds like it'd work.

Effectively what you're saying is _dont_ turn any docker droplets into kpack droplets. Instead, require repushing. 

This way, we'll know that kpack droplets have usable start commands and the migration won't break any apps that just happen to work in the current, weird state. Otherwise, switching to use `started_command` could break apps on migration because we weren't using them before and were likely saving the wrong data.

Reply:
> Otherwise, switching to use started_command could break apps on migration because we weren't using them before and were likely saving the wrong data.

Yep, when I tried doing this in this branch `dora` would just run `bundle` and install gems instead of `bundle rackup`.

Reply:
I rebased on master and removed the migration as part of that. It's no longer present in this commit:
https://github.com/cloudfoundry/cloud_controller_ng/pull/1917/commits/87d0369f72ccd411b6ce2551d42105b1d9e6a48a